File: Sliding Window/3. Longest Substring Without Repeating Characters/3. Longest Substring Without Repeating Characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
  let l = 0;
  let sett = new Set();
  let longest = 0;

  for (let r = 0; r < s.length; r++) {
    while (sett.has(s[r])) {
      sett.delete(s[l]);
      l++;
    }

    sett.add(s[r]);
    longest = Math.max(longest, sett.size);
  }

  return longest;
};

// Time: O(n)
// Space: O(n)

File: Sliding Window/3. Longest Substring Without Repeating Characters/3. Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l = 0
        longest = 0
        sett = set()

        for r in range(len(s)):
            while s[r] in sett:
                sett.remove(s[l])
                l += 1

            sett.add(s[r])
            longest = max(longest, len(sett))
        
        return longest
            
# Time: O(n)
# Space: O(n)